File: src/main/java/pl/training/workshop/m6/s12_onemany/step2/RefundService.java
package pl.training.workshop.m6.s12_onemany.step2;

import java.time.LocalDateTime;
import java.util.List;

import pl.training.workshop.m6.s12_onemany.TicketData;
import pl.training.workshop.shared.Money;

/**
 * Krok 2: Replace One/Many Distinctions with Composite - jedna metoda refund(Refundable).
 * Stare metody zostają jako cienkie delegacje, dopóki klienci nie przejdą na nowy kontrakt.
 */
public final class RefundService {
    private static final Money FEE = Money.of("3.00");

    public Money refund(Refundable refundable, LocalDateTime now) {
        return refundable.refundableAmount(now).minus(FEE).max(Money.ZERO);
    }

    /** @deprecated użyj {@link #refund(Refundable, LocalDateTime)} z {@link SingleTicket}. */
    @Deprecated
    public Money refund(TicketData ticket, LocalDateTime now) {
        return refund(new SingleTicket(ticket), now);
    }

    /** @deprecated użyj {@link #refund(Refundable, LocalDateTime)} z {@link TicketGroup}. */
    @Deprecated
    public Money refundAll(List<TicketData> tickets, LocalDateTime now) {
        return refund(TicketGroup.of(tickets), now);
    }
}

File: typescript/src/workshop/m6/s12_onemany/step2/RefundService.ts
import { Money } from '../../../shared/Money.js';
import type { LocalDateTime } from '../../../shared/time.js';
import type { TicketData } from '../TicketData.js';
import type { Refundable } from './Refundable.js';
import { SingleTicket } from './SingleTicket.js';
import { TicketGroup } from './TicketGroup.js';

/**
 * Krok 2: Replace One/Many Distinctions with Composite - jedna metoda refund(Refundable).
 * Stare metody zostają jako cienkie delegacje, dopóki klienci nie przejdą na nowy kontrakt.
 */
export class RefundService {
  private static readonly FEE = Money.of('3.00');

  refund(refundable: Refundable, now: LocalDateTime): Money {
    return refundable.refundableAmount(now).minus(RefundService.FEE).max(Money.ZERO);
  }

  /**
   * Dawne refund(ticket, now) - TypeScript nie ma przeciążeń, więc stara metoda dostaje osobną nazwę.
   * @deprecated użyj {@link RefundService.refund} z {@link SingleTicket}.
   */
  refundTicket(ticket: TicketData, now: LocalDateTime): Money {
    return this.refund(new SingleTicket(ticket), now);
  }

  /** @deprecated użyj {@link RefundService.refund} z {@link TicketGroup}. */
  refundAll(tickets: readonly TicketData[], now: LocalDateTime): Money {
    return this.refund(TicketGroup.of(tickets), now);
  }
}
